core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java

```diff
@@ -0,0 +1,109 @@
+ */
+package io.trino.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import io.trino.Session;
+import io.trino.metadata.Metadata;
+import io.trino.metadata.ResolvedFunction;
+import io.trino.operator.scalar.ArrayDistinctFunction;
+import io.trino.operator.scalar.ArraySortFunction;
+import io.trino.spi.type.Type;
+import io.trino.sql.PlannerContext;
+import io.trino.sql.planner.FunctionCallBuilder;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.ExpressionTreeRewriter;
+import io.trino.sql.tree.FunctionCall;
+import io.trino.sql.tree.QualifiedName;
+import io.trino.sql.tree.SymbolReference;
+
+import java.util.List;
+import java.util.Set;
+
+public class ArraySortAfterArrayDistinct
+        extends ExpressionRewriteRuleSet
+{
+    public ArraySortAfterArrayDistinct(PlannerContext plannerContext)
+    {
+        super((expression, context) -> rewrite(expression, context, plannerContext.getMetadata()));
+    }
+
+    @Override
+    public Set<Rule<?>> rules()
+    {
+        return ImmutableSet.of(
+                projectExpressionRewrite(),
+                filterExpressionRewrite(),
+                joinExpressionRewrite(),
+                valuesExpressionRewrite(),
+                patternRecognitionExpressionRewrite());
+    }
+
+    private static Expression rewrite(Expression expression, Rule.Context context, Metadata metadata)
+    {
+        if (expression instanceof SymbolReference) {
+            return expression;
+        }
+        Session session = context.getSession();
+        return ExpressionTreeRewriter.rewriteWith(new Visitor(metadata, session), expression);
+    }
+
+    private static class Visitor
+            extends io.trino.sql.tree.ExpressionRewriter<Void>
+    {
+        private final Metadata metadata;
+        private final Session session;
+
+        public Visitor(Metadata metadata, Session session)
+        {
+            this.metadata = metadata;
+            this.session = session;
+        }
+
+        @Override
+        public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            FunctionCall rewritten = treeRewriter.defaultRewrite(node, context);
+            if (metadata.decodeFunction(rewritten.getName()).getSignature().getName().equals(ArrayDistinctFunction.NAME) &&
+                    rewritten.getArguments().size() == 1 &&
+                    rewritten.getArguments().get(0) instanceof FunctionCall &&
+                    metadata.decodeFunction(((FunctionCall) rewritten.getArguments().get(0)).getName()).getSignature().getName().equals(ArraySortFunction.NAME)) {
+                FunctionCall arraySortFunctionCall = (FunctionCall) rewritten.getArguments().get(0);
+                ResolvedFunction arraySortResolvedFunction = metadata.decodeFunction(arraySortFunctionCall.getName());
```

Comment:
Please try to structure the code so as to call `metadata.decodeFunction` once for a Functioncall.